ergo-chain-sync/src/cache/redis.rs

```diff
@@ -0,0 +1,206 @@
+        let block_height_key = concat_string!(block_id_hex, ":h");
+        let block_transactions_key = concat_string!(block_id_hex, ":t");
+
+        let mut pipe = redis::pipe();
+        // SET {block_id_hex}:p {parent_id_hex}
+        let () = pipe
+            .cmd("SET")
+            .arg(&parent_id_key)
+            .arg(&parent_id_hex)
+            .ignore()
+            // SET {block_id_hex}:h {block.height}
+            .cmd("SET")
+            .arg(&block_height_key)
+            .arg(block.height)
+            .query_async(&mut conn)
+            .await
+            .unwrap();
+
+        let mut pipe = redis::pipe();
+        pipe.cmd("MULTI").ignore();
+        for tx in block.transactions {
+            let tx_json = serde_json::to_string(&tx).unwrap();
+            let tx_id = String::from(tx.id().0);
+
+            // SADD {block_id_hex}:t {tx_id}
+            pipe.cmd("SADD")
+                .arg(&block_transactions_key)
+                .arg(&tx_id)
+                .ignore()
+                // SET {tx_id} {tx_json}
+                .cmd("SET")
+                .arg(&tx_id)
+                .arg(&tx_json)
+                .ignore();
+        }
+        let () = pipe.cmd("EXEC").query_async(&mut conn).await.unwrap();
+
+        // DEL best_block
+        let _: () = pipe
+            .cmd("DEL")
+            .arg(BEST_BLOCK)
+            .ignore()
+            // RPUSH best_block {block_id_hex} {parent_id_hex}
+            .cmd("RPUSH")
+            .arg(BEST_BLOCK)
+            .arg(&block_id_hex)
+            .arg(&parent_id_hex)
+            .query_async(&mut conn)
+            .await
+            .unwrap();
+    }
+
+    async fn exists(&mut self, block_id: BlockId) -> bool {
+        let mut conn = self.pool.get().await.unwrap();
+        let block_id_hex = String::from(block_id.0);
+        let block_height_key = concat_string!(block_id_hex, ":h");
+        let exists: bool = cmd("EXISTS")
+            .arg(&block_height_key)
+            .query_async(&mut conn)
+            .await
+            .unwrap();
+        exists
+    }
+
+    async fn get_best_block(&mut self) -> Option<BlockRecord> {
+        let mut conn = self.pool.get().await.unwrap();
+        let block_id_hex: Vec<String> = cmd("LRANGE")
+            .arg(BEST_BLOCK)
+            .arg(0)
+            .arg(0)
+            .query_async(&mut conn)
+            .await
+            .unwrap();
+        if !block_id_hex.is_empty() {
+            let digest32 = Digest32::try_from(block_id_hex[0].clone()).unwrap();
+            let id = BlockId(digest32);
+            let height: u32 = cmd("GET")
+                .arg(&concat_string!(block_id_hex[0], ":h"))
+                .query_async(&mut conn)
+                .await
+                .unwrap();
+            Some(BlockRecord { id, height })
+        } else {
+            None
+        }
+    }
+
+    async fn take_best_block(&mut self) -> Option<Block> {
```

Comment:
This operation is worth making atomic as well.

Reply:
I've looked into it and I don't think it can be done. The reason is that we need to pull down some results from Redis to the Rust-side (e.g. `transaction_ids` needed for `MGET` command), and use that information for subsequent commands. I've looked into Redis scripting that allows you perform more logic server side. But it can't apply for this situation.

In https://redis.io/docs/manual/programmability/eval-intro/:
<img width="759" alt="image" src="https://user-images.githubusercontent.com/791016/204988405-2e2033e5-e8da-45f2-9904-b4de81bb8dfd.png">

That last sentence is the key part.